modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/components/FieldSelectModalContent.tsx

```diff
@@ -0,0 +1,364 @@
+import {getFields} from '../api';
+import {IField} from '../utils/types';
+import Search from './Search';
+
+interface IFieldTreeItem extends IField {
+	children?: IFieldTreeItem[];
+	query?: string;
+	selected?: boolean;
+}
+
+function visit(fields: Array<IFieldTreeItem>, callback: Function) {
+	fields.forEach((field) => {
+		callback(field);
+
+		if (field.children) {
+			visit(field.children, callback);
+		}
+	});
+}
+
+const initializeFields = ({
+	fields: initialFields,
+	selectedFields,
+}: {
+	fields: IField[];
+	selectedFields: Array<IField>;
+}): [Set<React.Key>, Array<IFieldTreeItem>] => {
+	const selectedKeys = new Set<React.Key>();
+	const fields: IFieldTreeItem[] = Array.from(initialFields);
+
+	visit(fields, (field: IFieldTreeItem) => {
+		const selectedField = selectedFields.find(
+			(selectedField) => selectedField.name === field.name
+		);
+
+		if (selectedField) {
+			selectedKeys.add(selectedField.name);
+		}
+
+		field.id = field.name;
+	});
+
+	return [selectedKeys, fields];
+};
+
+function filterFields({
+	fields,
+	onFilter,
+	onMatch,
+	query,
+}: {
+	fields: Array<IFieldTreeItem>;
+	onFilter?: (field: IFieldTreeItem) => void;
+	onMatch?: (field: IFieldTreeItem) => void;
+	query: string;
+}) {
+	const filteredItems: Array<IFieldTreeItem> = [];
+	const regexp = new RegExp(query, 'i');
+
+	fields.forEach((field) => {
+		const match = field.label ? regexp.test(field.label) : false;
+
+		const filteredChildren = field.children?.length
+			? filterFields({fields: field.children, onFilter, onMatch, query})
+			: [];
+
+		if (match || (field.children?.length && filteredChildren.length)) {
+			filteredItems.push({
+				...field,
+				children: filteredChildren,
+				query,
+			});
+
+			if (onFilter) {
+				onFilter(field);
+			}
+		}
+
+		if (match && onMatch) {
+			onMatch(field);
+		}
+	});
+
+	return filteredItems;
+}
+
+function applyFilter({
+	fields,
+	query,
+}: {fields?: Array<IFieldTreeItem>; query?: string} = {}) {
+	if (!query || !fields) {
+		return {
+			counter: 0,
+			filteredItems: fields ?? [],
+			filteredKeys: [],
+		};
+	}
+
+	let counter = 0;
+	const filteredKeys: Array<React.Key> = [];
+	const filteredItems = filterFields({
+		fields,
+		onFilter: ({id}: IFieldTreeItem) => {
+			if (id) {
+				filteredKeys.push(id);
+			}
+		},
+		onMatch: () => counter++,
+		query,
+	});
+
+	return {
+		counter,
+		filteredItems,
+		filteredKeys,
+	};
+}
+
+const Highlight = ({query, text}: {query?: string; text?: string}) => {
+	if (!query || !text) {
+		return <>{text ?? ''}</>;
+	}
+
+	const indexMatch = text.search(RegExp(query, 'i'));
+
+	return indexMatch > -1 ? (
+		<>
+			{text.substring(0, indexMatch)}
+			<mark className="bg-transparent border-0 font-weight-bold p-0 shadow-none">
+				{text.substring(indexMatch, indexMatch + query.length)}
+			</mark>
+			{text.substring(indexMatch + query.length)}
+		</>
+	) : (
+		<>{text}</>
+	);
+};
+
+const FieldSelectModalContent = ({
```

Comment:
There is a high amount of code duplication with this component and `AddFieldsModalContent`. That component served as a starting point, but I took out saving logic. The goal was to make this component only for selection, reusable, with saving itself done in parent contexts.

I left in field specifics, as fields are the only hierarchical select in DSM scope. This could be made more abstract, something like a `TreeviewSelectModalContent`, or even `TreeviewSelect`, but that would require more extensive API and taking out many DSM specifics. That may be a future task.

The intent is to migrate `AddFieldsModalContent` to use `FieldSelectModalContent`. This can be in a followup, this PR is already big enough.